servicetalk-opentracing-zipkin-publisher/src/main/java/io/servicetalk/opentracing/zipkin/publisher/reporter/HttpReporter.java

```diff
@@ -0,0 +1,287 @@
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import zipkin2.CheckResult;
+import zipkin2.Component;
+import zipkin2.Span;
+import zipkin2.codec.SpanBytesEncoder;
+import zipkin2.reporter.Reporter;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+import javax.annotation.Nonnull;
+
+import static io.servicetalk.concurrent.api.AsyncCloseables.newCompositeCloseable;
+import static io.servicetalk.concurrent.api.AsyncCloseables.toAsyncCloseable;
+import static io.servicetalk.concurrent.api.BufferStrategies.forCountOrTime;
+import static io.servicetalk.concurrent.api.Completable.completed;
+import static io.servicetalk.concurrent.api.Processors.newCompletableProcessor;
+import static io.servicetalk.concurrent.api.Processors.newPublisherProcessorDropHeadOnOverflow;
+import static io.servicetalk.concurrent.api.SourceAdapters.fromSource;
+import static io.servicetalk.concurrent.api.SourceAdapters.toSource;
+import static io.servicetalk.concurrent.internal.FutureUtils.awaitTermination;
+import static io.servicetalk.http.api.CharSequences.newAsciiString;
+import static io.servicetalk.http.api.HttpHeaderNames.CONTENT_TYPE;
+import static io.servicetalk.http.api.HttpHeaderValues.APPLICATION_JSON;
+import static java.time.Duration.ofSeconds;
+import static java.util.Objects.requireNonNull;
+import static zipkin2.CheckResult.OK;
+import static zipkin2.CheckResult.failed;
+
+/**
+ * A {@link Span} {@link Reporter} that will publish to an HTTP endpoint with a configurable encoding {@link Codec}.
+ */
+public final class HttpReporter extends Component implements Reporter<Span>, AsyncCloseable {
+    private static final Logger LOGGER = LoggerFactory.getLogger(HttpReporter.class);
+    static final String V1_PATH = "/api/v1/spans";
+    static final String V2_PATH = "/api/v2/spans";
+    static final CharSequence THRIFT_CONTENT_TYPE = newAsciiString("application/x-thrift");
+    static final CharSequence PROTO_CONTENT_TYPE = newAsciiString("application/protobuf");
+
+    private final PublisherSource.Processor<Span, Span> buffer;
+    private final CompositeCloseable closeable;
+
+    private volatile boolean closeInitiated;
+
+    private HttpReporter(final Builder builder) {
+        closeable = newCompositeCloseable();
+        final HttpClient client = closeable.append(builder.clientBuilder.build());
+        try {
+            buffer = initReporter(builder, client);
+        } catch (Throwable t) {
+            try {
+                closeable.close();
+            } catch (Exception e) {
+                LOGGER.error("Failed to close the client.", e);
+            }
+            throw t;
+        }
+    }
+
+    @Override
+    public CheckResult check() {
+        return closeInitiated ? OK : failed(new IllegalStateException("Reporter is closed."));
+    }
+
+    @Override
+    public void report(final Span span) {
+        if (closeInitiated) {
```

Comment:
since this is best effort, and in practice may not be triggered, does it make sense to drop this this check to avoid providing a false sense of security (e.g. it triggers in tests, but not in real world)?

Reply:
> and in practice may not be triggered,

I think you are saying that in practice it won't be triggered when the usage is as expected, i.e. no spans are reported after the reporter is closed.  I don't think we can assume expected usage. This is a trade-off between being extremely defensive (making sure no spans can be reported after closure, even when they are concurrent) at the cost of complexity and performance or a best-effort check to provide visibility when an inadvertent misuse has happened, like the same `HttpReporter` instance is shared across multiple `ZipkinPublisher`s.

This trade-off seems reasonable to me.

Reply:
To clarify "best effort" was meant to highlight even if it isn't used as expected this code may not provide the desired result (e.g. throwing an error). It depends upon a happens-before relationship between closing and reporting which may or may-not exist. I'm suggesting just removing the check because it isn't required from the API contract, there is no guarantee it will throw (e.g. silently dropping data may happen anyways and should be tolerated, users can't rely upon this class detecting use-after-close, which may not be clear from code inspection or unit tests that have different lifetimes/barriers), and it introduces code into the hot path.

Reply:
I understood your suggestion. 

I think you are overlooking that this field is `volatile`?